src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -422,6 +439,29 @@ private static RecursivePatternSyntax RecursivePattern(SubpatternSyntax subpatte
                 commonReceiver = GetInnermostReceiver(left, lastName, static (identifierName, lastName) => identifierName != lastName);
             }
 
+            var iSymbol = model.GetSymbolInfo(leftNames[^1]).Symbol;
+            if (iSymbol != null && iSymbol.Equals(model.GetSymbolInfo(rightNames[^1]).Symbol))
+            {
+                var leftIsNullCheck = IsNullCheck(leftTarget.Parent);
+                var rightIsNullCheck = IsNullCheck(rightTarget.Parent);
+
+                if (leftIsNullCheck)
+                {
+                    lastName = rightNames[^1];
+                    commonReceiver = GetInnermostReceiver(right, lastName, static (identifierName, lastName) => identifierName != lastName);
+                    rightNames.Clip(rightNames.Count - 1);
+                    return (commonReceiver ?? ThisExpression(), ImmutableArray<IdentifierNameSyntax>.Empty, rightNames.ToImmutable());
+                }
+
+                if (rightIsNullCheck)
+                {
+                    lastName = leftNames[^1];
+                    commonReceiver = GetInnermostReceiver(left, lastName, static (identifierName, lastName) => identifierName != lastName);
+                    leftNames.Clip(leftNames.Count - 1);
+                    return (commonReceiver ?? ThisExpression(), leftNames.ToImmutable(), ImmutableArray<IdentifierNameSyntax>.Empty);
+                }
```

Comment:
all of this needs docs.

Reply:
The comments hopefully explains it better, see if it's enough

Reply:
this de feels like we should have a helper that both blocks share.